File: python/satellite/satellite.py
def tree_from_traversals(preorder, inorder):
    if len(preorder) != len(inorder):
        raise ValueError('traversals must have the same length')
    if set(preorder) != set(inorder):
        raise ValueError('traversals must have the same elements')
    if len(preorder) != len(set(preorder)):
        raise ValueError('traversals must contain unique items')
    return build_tree(preorder, inorder)


def build_tree(preorder, inorder):
    '''
    I broke the recursive part into a separate function
    only because the initial assertions only have to be
    checked once.
    '''
    if len(preorder) == 0:
        return {}
    value = preorder[0]
    idx = inorder.index(value)
    return {
        'v': value,
        'l': build_tree(preorder[1:1+idx], inorder[:idx]),
        'r': build_tree(preorder[idx+1:], inorder[idx+1:]),
    }

File: elixir/satellite/lib/satellite.ex
defmodule Satellite do
  @typedoc """
  A tree, which can be empty, or made from a left branch, a node and a right branch
  """
  @type tree :: {} | {tree, any, tree}

  @doc """
  Build a tree from the elements given in a pre-order and in-order style
  """
  @spec build_tree(preorder :: [any], inorder :: [any]) :: {:ok, tree} | {:error, String.t()}
  def build_tree(preorder, inorder) do
    len = length(preorder)
    cond do
      len != length(inorder) ->
        {:error, "traversals must have the same length"}

      Enum.any?(preorder, &(&1 not in inorder)) ->
        {:error, "traversals must have the same elements"}

      len != MapSet.size(MapSet.new(preorder)) ->
        {:error, "traversals must contain unique items"}

      true ->
        {:ok, do_build_tree(preorder, inorder)}
    end
  end

  defp do_build_tree([], []), do: {}

  defp do_build_tree(preorder, inorder) do
    [root | preorder_rest] = preorder
    n = Enum.find_index(inorder, &(&1 == root))

    preorder_left = Enum.take(preorder_rest, n)
    preorder_right = Enum.drop(preorder_rest, n)

    inorder_left = Enum.take(inorder, n)
    inorder_right = Enum.drop(inorder, n + 1)

    {
      do_build_tree(preorder_left, inorder_left),
      root,
      do_build_tree(preorder_right, inorder_right)
    }
  end
end
